src/diary.cpp

```diff
@@ -0,0 +1,791 @@
+    } else {
+
+        bool flag = true;
+        for( const proficiency_id &elem : currpage->known_profs ) {
+
+            if( std::find( prevpage->known_profs.begin(), prevpage->known_profs.end(),
+                           elem ) == prevpage->known_profs.end() ) {
+                if( flag ) {
+                    add_to_change_list( _( "Proficiencies: " ) );
+                    flag = false;
+                }
+                const proficiency &p = elem.obj();
+                add_to_change_list( p.name(), p.description() );
+            }
+        }
+        if( !flag ) {
+            add_to_change_list( " " );
+        }
+    }
+}
+
+std::vector<std::string> diary::get_change_list()
+{
+    if( !change_list.empty() ) {
+        return change_list;
+    }
+    if( !pages.empty() ) {
+        stat_changes();
+        skill_changes();
+        prof_changes();
+        trait_changes();
+        bionic_changes();
+        spell_changes();
+        mission_changes();
+        kill_changes();
+    }
+    return change_list;
+}
+
+std::map<int, std::string> diary::get_desc_map()
+{
+    if( !desc_map.empty() ) {
+        return desc_map;
+    } else {
+        get_change_list();
+        return desc_map;
+    }
+}
+
+
+std::string diary::get_page_text()
+{
+
+    if( !pages.empty() ) {
+        return get_page_ptr()->m_text;
+    }
+    return "";
+}
+
+std::string diary::get_head_text()
+{
+
+    if( !pages.empty() ) {
+        const diary_page *prevpageptr = get_page_ptr( -1 );
+        const time_point prev_turn = ( prevpageptr != nullptr ) ? prevpageptr->turn : calendar::turn_zero;
+        const time_duration turn_diff = get_page_ptr()->turn - prev_turn;
+        const int days = to_days<int>( turn_diff );
+        const int hours = to_hours<int>( turn_diff ) % 24;
+        const int minutes = to_minutes<int>( turn_diff ) % 60;
+        std::string headtext = string_format( _( "Entry: %d/%d, %s, %s" ),
+                                              opend_page + 1, pages.size(),
+                                              to_string( get_page_ptr()->turn ),
+                                              ( opend_page != 0 ) ? string_format( _( "%s%s%d minutes since last entry" ),
+                                                      ( days > 0 ) ? string_format( _( "%d days, " ), days ) : "",
+                                                      ( hours > 0 ) ? string_format( _( "%d hours, " ), hours ) : "",
+                                                      minutes ) : "" );
+
+        return headtext;
+    }
+    return "";
+}
+
+void diary::death_entry()
+{
+    bool lasttime = query_yn( _( "Open diary for the last time?" ) );
+    if( lasttime ) {
+        show_diary_ui( this );
+    }
+    export_to_txt( true );
+}
+
+diary::diary()
+{
+    owner = get_avatar().name;
+}
+void diary::set_page_text( std::string text )
+{
+    get_page_ptr()->m_text = text;
+}
+
+void diary::new_page()
+{
+    diary_page *page = new diary_page();
```

Comment:
You allocated this `diary_page` object but didn't seem to be freeing it anywhere.

Reply:
Thanks for the feedback. I changed pages to use unique pointer and change the rest accordingly. Or is there a better way to do it?  https://github.com/CleverRaven/Cataclysm-DDA/blob/32a509ad18853aceb31f8b8862f2946020fec6b3/src/diary.h#L72 
https://github.com/CleverRaven/Cataclysm-DDA/blob/32a509ad18853aceb31f8b8862f2946020fec6b3/src/diary.cpp#L625